redun/executors/k8s.py

```diff
@@ -0,0 +1,998 @@
+
+
+def is_array_job_name(job_name: str) -> bool:
+    "Returns true if the job name looks like an array job"
+    return job_name.endswith(f"-{ARRAY_JOB_SUFFIX}")
+
+
+def get_hash_from_job_name(job_name: str) -> Optional[str]:
+    """
+    Returns the job/task eval_hash that corresponds with a particular job name
+    on K8S.
+    """
+    # Remove array job suffix, if present.
+    array_suffix = "-" + ARRAY_JOB_SUFFIX
+    if job_name.endswith(array_suffix):
+        job_name = job_name[: -len(array_suffix)]
+
+    # It's possible we found jobs that are unrelated to the this work based off
+    # the job_name_prefix matching when fetching in get_jobs. These jobs will
+    # not have hashes so we can ignore them. For a concrete example of this,
+    # see:
+    #
+    #   https://insitro.atlassian.net/browse/DE-2632
+    #
+    # where a headnode job is running but has no hash so we don't want to
+    # interact with that job here. If we don't find a match, consider this a
+    # case of the above where we matched unrelated jobs and return None to let
+    # callers know this is the case.
+    match = re.match(".*-(?P<hash>[^-]+)", job_name)
+    if match:
+        return match["hash"]
+
+    return None
+
+
+def get_k8s_job_options(job_options: dict) -> dict:
+    """
+    Returns K8S-specific job options from general job options.
+    """
+    keys = ["memory", "vcpus", "k8s_labels", "retries", "timeout"]
+    return {key: job_options[key] for key in keys if key in job_options}
+
+
+def submit_task(
+    image: str,
+    namespace: str,
+    s3_scratch_prefix: str,
+    job: Job,
+    a_task: Task,
+    args: Tuple = (),
+    kwargs: Dict[str, Any] = {},
+    job_options: dict = {},
+    array_uuid: Optional[str] = None,
+    array_size: int = 0,
+    code_file: Optional[File] = None,
+) -> kubernetes.client.V1Job:
+    """
+    Submit a redun Task to K8S.
+    """
+    command = get_oneshot_command(
+        s3_scratch_prefix,
+        job,
+        a_task,
+        args,
+        kwargs,
+        job_options=job_options,
+        code_file=code_file,
+        array_uuid=array_uuid,
+    )
+
+    if array_uuid:
+        job_hash = array_uuid
+    else:
+        assert job.eval_hash
+        job_hash = job.eval_hash
+
+    # Submit to K8S
+    job_name = get_k8s_job_name(
+        job_options.get("job_name_prefix", "k8s-job"),
+        job_hash,
+        array=bool(array_size),
+    )
+    result = k8s_submit(
+        command,
+        image=image,
+        namespace=namespace,
+        job_name=job_name,
+        array_size=array_size,
+        **get_k8s_job_options(job_options),
+    )
+
+    return result
+
+
+def submit_command(
+    image: str,
+    namespace: str,
+    scratch_prefix: str,
+    job: Job,
+    command: str,
+    job_options: dict = {},
+) -> kubernetes.client.V1Job:
+    """
+    Submit a shell command to K8S
+    """
+    shell_command = get_script_task_command(
```

Comment:
@dakoner I have now generalized file staging beyond just s3. That allows us to generalize this shell command generation to work with any object storage, not just s3 as it was before.